db/version_set.cc

```diff
@@ -4909,29 +4909,57 @@ uint64_t VersionSet::ApproximateSize(Version* v, const Slice& start,
     assert(files_brief.num_files > 0);
 
     // identify the file position for starting key
-    const uint64_t idx_start = FindFileInRange(
-        v->cfd_->internal_comparator(), files_brief, start,
-        /*start=*/0, static_cast<uint32_t>(files_brief.num_files - 1));
-    assert(idx_start < files_brief.num_files);
-
-    // scan all files from the starting position until the ending position
-    // inferred from the sorted order
-    for (uint64_t i = idx_start; i < files_brief.num_files; i++) {
-      uint64_t val;
-      val = ApproximateSize(v, files_brief.files[i], end, caller);
-      if (!val) {
-        // the files after this will not have the range
-        break;
+    const int idx_start =
+        FindFileInRange(v->cfd_->internal_comparator(), files_brief, start, 0,
+                        static_cast<uint32_t>(files_brief.num_files - 1));
+    assert(static_cast<size_t>(idx_start) < files_brief.num_files);
+    const int idx_end = FindFileInRange(
+        v->cfd_->internal_comparator(), files_brief, end, idx_start,
+        static_cast<uint32_t>(files_brief.num_files - 1));
```

Comment:
It's common that the range is really small so that idx_start and ids_end fall into the same file. We may consider to avoid a full binary search in that case.

Reply:
But without the search - we can't be sure. Also this is pretty cheap.
Having said that - I think I have an idea on how to optimize it a bit.